feat(chain): Improved `changes` RPC

frol opened the pull request:
*Resolves #2034 and #2048*

The `changes` RPC API was broken implementation-wise (#2048), and design-wise (#2034).

This version has meaningful API design for all the exposed data, and it is also tested better. This PR is massive since, initially, we missed quite a point of exposing deserialized internal data (like account info, access key used to be returned as borsh serialized blob, which is useless for the API user as they don't have easy access to the schema of those structures).

I will address #2129 in a separate PR.

## API

`EXPERIMENTAL_changes` RPC endpoint accepts the following types of requests:

```js
{
    "block_id": /* block hash | block height | omit it if you want to use `finality` */,
    "finality": /* "latest" | "final" | "near-final" | omit it if you want to se `block_id` */,
    "changes_type": /* "account_changes" | "single_access_key_changes" | "all_access_key_changes" | "code_changes" | "data_changes" */,
    /* extra fields based on `changes_type` */
}
```

the extra fields based on `changes_type` are the following:

* ` "account_changes"`:
    * `account_id`
* `"single_access_key_changes"`:
    * `account_id`
    * `access_key_pk` (public key)
* `"all_access_key_changes"`:
    * `account_id`
* `"code_changes`:
    * `account_id`
* `"data_changes"`:
    * `account_id`
    * `key_prefix_base64` (base64-encoded string of the key-prefix data saved from a smart contract)

### Example of Account changes

When an account gets modified (created, updated, removed), we can query the block which contains the receipts for the transaction:

```js
http post http://localhost:3030 jsonrpc=2.0 id=dontcare \
    method=EXPERIMENTAL_changes \
    'params:={
        "block_id": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW"
        "changes_type": "account_changes",
        "account_id": "new-account"
    }'
```

we observe the following:

```js
{
    "block_hash": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW",
    "changes": [
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "8rPN9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTobgRv"
            },
            "type": "account_update",
            "change": {
                "account_id": "new-account",
                "amount": "0",
                "locked": "0",
                "code_hash": "11111111111111111111111111111111",
                "storage_usage": 100,
                "storage_paid_at": 9
            }
        }
    ]
}
```

If the account was deleted:

```js
{
    "block_hash": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW",
    "changes": [
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "8rPN9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTobgRv"
            },
            "type": "account_deletion",
            "change": {
                "account_id": "new-account"
            }
        }
    ]
}
```

### Example of Data changes

When some key gets modified (created, updated, removed), we can query the block which contains the receipts for the transaction:

```js
http post http://localhost:3030 jsonrpc=2.0 id=dontcare \
    method=EXPERIMENTAL_changes \
    'params:={
        "block_id": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW"
        "changes_type": "data_changes",
        "account_id": "new-account",
        "key_prefix_base64": "bXlfa2V5"  /* base64-encoded "my_key" */
    }'
```

we observe the following (assume the key was modified twice in a single block, and then removed):

```js
{
    "block_hash": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW",
    "changes": [
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "111N9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTob111",
            },
            "type": "data_update",
            "change": {
                "account_id": "new-account",
                "key_base64": "bXlfa2V5",  /* base64-encoded "my_key" */
                "value_base64":  "bXlfdmFsdWVfMQ==" /* base64-encoded "my_value_1" */,
            }
        },
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "222N9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTob222"
            },
            "type": "data_update",
            "change": {
                "account_id": "new-account",
                "key_base64": "bXlfa2V5",  /* base64-encoded "my_key" */
                "value_base64": "bXlfdmFsdWVfMg==" /* base64-encoded "my_value_2" */
            }
        },
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "333N9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTob333"
            },
            "type": "data_deletion",
            "change": {
                "account_id": "new-account",
                "key_base64": "bXlfa2V5"  /* base64-encoded "my_key" */
            }
        }
    ]
}
```

## Test plan

I did not succeed in writing Rust tests as we used a mocked runtime there. Thus, I had extended end-to-end tests (pytest) with:
* Test for account changes on account creation
* Test for access key changes on account creation and access key removal
* Test for code changes
* Test for several transactions on the same block/chunk

frol commented on chain/jsonrpc/tests/rpc_transactions.rs:
```diff
@@ -133,45 +133,40 @@ fn test_expired_tx() {
-                                            .map(|_| ()),
-                                    );
-                                }
+                actix::spawn(view_client.send(GetBlock::latest()).then(move |res| {
```
It is just a change in code formatting due to `GetBlock::latest()` change.

frol commented on chain/network/src/types.rs:
```diff
@@ -1198,7 +1196,7 @@ pub enum NetworkViewClientMessages {
     TxStatusResponse(FinalExecutionOutcomeView),
     /// General query
-    Query { query_id: String, block_id: MaybeBlockId, request: QueryRequest, finality: Finality },
+    Query { query_id: String, block_checkpoint: BlockCheckpoint, request: QueryRequest },
```
It did not make sense for `block_id` and `finality` to co-exist. I had to make these changes in this PR as I was extending `changes` API (I should have created a separate PR, but it was hard to decouple the changes).

frol commented on core/primitives/Cargo.toml:
```diff
@@ -6,11 +6,11 @@ edition = "2018"
 
 [dependencies]
 regex = "1"
-bincode = { version = "1.0", features = ["i128"] }
```
It is not used anymore.

frol commented on core/primitives/Cargo.toml:
```diff
@@ -6,11 +6,11 @@ edition = "2018"
 base64 = "0.11"
 byteorder = "1.2"
 chrono = { version = "0.4.4", features = ["serde"] }
+derive_more = "0.99.3"
```
It helped to clean up some duplicate code around `AsRef` / `Into` / etc common implementations.

frol commented on core/primitives/src/serialize.rs:
```diff
@@ -78,34 +43,6 @@ pub trait BaseDecode: for<'a> TryFrom<&'a [u8], Error = Box<dyn std::error::Erro
     }
 }
 
-macro_rules! pretty_bytes_serialization {
```
I decided to be more explicit and use `_base64` suffix on the fields serialized with base64 encoding, which requires serde renaming anyway, so I rolled back this macro implementation and replaced it with appropriate field attributes.

frol commented on core/primitives/src/serialize.rs:
```diff
@@ -1,50 +1,15 @@
-pub type DecodeResult<T> = Result<T, io::Error>;
-
-// encode a type to byte array
-pub trait Encode {
```
Cleaned up dead code

frol commented on core/primitives/src/views.rs:
```diff
@@ -1041,19 +1059,122 @@ pub struct GasPriceView {
 
-/// Different types of finality.
-#[derive(Serialize, Deserialize, BorshSerialize, BorshDeserialize, Clone, Debug, PartialEq, Eq)]
-pub enum Finality {
```
Moved to `types.rs`

frol commented on near/tests/stake_nodes.rs:
```diff
@@ -445,38 +441,28 @@ fn test_inflation() {
-                        },
-                    ),
-                );
+                actix::spawn(test_nodes[0].view_client.send(GetBlock::latest()).then(move |res| {
```
Nothing except `GetBlock::latest()` has changed in this block.

frol commented on pytest/tests/sanity/rpc_key_value_changes.py:
```diff
@@ -3,98 +3,171 @@
+            }
+        }
+    ]
+}
```
Example of the RPC response for account changes.

frol commented on pytest/tests/sanity/rpc_key_value_changes.py:
```diff
@@ -3,98 +3,171 @@
         }
-    )
-    assert not deepdiff.DeepDiff(changes_from_another_node, changes), "query same changes gives different result"
+    ]
```
Example of changes RPC response when two transactions in a single block cause a change in the same key.

frol commented on core/primitives/src/utils.rs:
```diff
@@ -35,89 +35,250 @@ pub mod col {
-    key
-}
+#[derive(derive_more::AsRef, derive_more::Into)]
+struct KeyForColumnAccountId(Vec<u8>);
```
I had to refactor this file as I was adding a few more functions and already felt like loosing track of them, so I essentially grouped the functions into namespaces. /cc @mikhailOK it seems that you have been working on these before. Please, review the changes.

frol commented on chain/chain/src/store.rs:
```diff
@@ -923,50 +923,175 @@ impl ChainStoreAccess for ChainStore {
         state_changes_request: &StateChangesRequest,
     ) -> Result<StateChanges, Error> {
         use near_primitives::utils;
+
```
FYI, the essential changes are in this method.

frol commented on chain/chain/src/store.rs:
```diff
@@ -923,50 +923,175 @@ impl ChainStoreAccess for ChainStore {
+
+                for change in changes_iter {
+                    let (_, state_changes) = change?;
+                    changes.extend(state_changes.into_iter().map(|(cause, state_change)| {
```
I have tried to express it in less code (e.g. I wanted to use `flat_map`, but we want to short-circuit when the iterator fails), and I have also tried to generalize the code to avoid semi-duplication in the matc branches, but it quickly becomes overcomplicated.

bowenwang1996 replied:
what is all the added code for?

bowenwang1996 commented on chain/jsonrpc/src/lib.rs:
```diff
@@ -529,32 +523,23 @@ impl JsonRpcHandler {
-        let (block_hash, state_changes_request) =
-            parse_params::<(CryptoHash, StateChangesRequest)>(params)?;
-        let block_hash_copy = block_hash.clone();
+        let RpcStateChangesRequest { block_checkpoint, state_changes_request } =
```
so this change is not backward compatible?

bowenwang1996 commented on core/primitives/src/types.rs:
```diff
@@ -54,38 +73,33 @@ pub struct AccountInfo {
-#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+/// This type is used to mark keys (arrays of bytes) that are queried from store.
+///
+/// NOTE: Currently, this type is only used in the client_view and RPC to be able to transparently
```
Do you mean view_client?

frol replied:
Fixed

nearmax commented on chain/chain/src/store.rs:
```diff
@@ -917,56 +917,187 @@ impl ChainStoreAccess for ChainStore {
+
+        let mut changes_iter =
+            self.store.iter_prefix_ser::<RawStateChangesList>(ColKeyValueChanges, &storage_key);
+        let mut exact_key_changes_iter;
```
Is this a lifetime hack? I am not sure it is needed here. We should be able to return `&mut changes_iter.filter_map(...` below.

frol replied:
Yes, it is. We cannot remove `&mut changes_iter.filter_map(...` as that is a temporary object created there with `filter_map`, so you cannot leak the reference.

nearmax replied:
Please add a comment.

nearmax commented on chain/chain/src/store.rs:
```diff
@@ -917,56 +917,187 @@ impl ChainStoreAccess for ChainStore {
+            self.store.iter_prefix_ser::<RawStateChangesList>(ColKeyValueChanges, &storage_key);
+        let mut exact_key_changes_iter;
+        let changes_iter: &mut dyn Iterator<Item = _> = match state_changes_request {
+            // These request types should have the key matching exactly
```
Is this a feature request? Why cannot we have partial key matching?

frol replied:
It does not seem sane to return the changes for all the accounts starting with `frol`, when the API request was to query the changes for AccountChanges or CodeChanges in account_id = frol. We may add separate types of requests with no problem, but this is a sane thing to keep these as it is now.

nearmax commented on chain/chain/src/store.rs:
```diff
@@ -917,56 +917,187 @@ impl ChainStoreAccess for ChainStore {
+            StateChangesRequest::AccountChanges { account_id, .. } => {
+                let mut changes = StateChanges::new();
+
+                for change in changes_iter {
```
We need a better name for this variable, e.g. changes_per_key

nearmax commented on chain/chain/src/store.rs:
```diff
@@ -917,56 +917,187 @@ impl ChainStoreAccess for ChainStore {
+            | StateChangesRequest::DataChanges { .. } => &mut changes_iter,
+        };
+
+        match state_changes_request {
```
I would extract this code into implementations of `From` trait for `StateChangeWithCause` from iterator of state changes, because this method is quite large.

frol replied:
Yeah, I will give another round of a try. My initial attempt was not successful (it failed to reduce the amount of code, and only made it worse, but I will see if StateChangeWithCause can save the day).

nearmax replied:
It might not reduce the amount of code but might split it into more encapsulated blocks, improving readability.

nearmax replied:
Also, when we have primitive type A constructible from primitive type B or its collection, then I think we should implement it as `From` trait and keep it next to the declaration of A or B. This seems to be a standard approach in std.

nearmax commented on core/primitives/src/types.rs:
```diff
@@ -253,6 +295,19 @@ pub enum BlockId {
 
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum BlockCheckpoint {
```
Block checkpoint might already have a meaning in our system, consider `BlockIdOrFinality`. CC @SkidanovAlex

frol replied:
Yeah, I have considered `BlockIdOrFinality`, but I did not like the verbosity of it. I am fine with renaming it, though.

frol replied:
Renamed to `BlockIdOrFinality`

frol commented on chain/chain/src/store.rs:
```diff
@@ -917,56 +917,187 @@ impl ChainStoreAccess for ChainStore {
+
         let mut storage_key = block_hash.as_ref().to_vec();
-        storage_key.extend(match state_changes_request {
+        let common_storage_key_prefix_len = storage_key.len();
```
We store data under a compound key: `block_hash + trie_key`, so when we query, we reverse the process of splitting the key by using simple slicing an array of bytes, essentially, extracting `trie_key`. Example: data changes are stored under a key:

```
block_hash + (col::ACCOUNT + account_id + ACCOUNT_DATA_SEPARATOR + user_specified_key)
```

Thus to query all the changes by a user-specified key prefix I do the following: query RocksDB for 

```
block_hash + (col::ACCOUNT + account_id + ACCOUNT_DATA_SEPARATOR + user_specified_key_prefix)
```

In the simplest case, to extract the full key I need to slice the RocksDB key by a length of

```
block_hash + (col::ACCOUNT + account_id + ACCOUNT_DATA_SEPARATOR)
```

In this implementation, I decoupled this process into two steps:

1. Split the block_hash, thus we are left working with a key that was used in trie
2. Parse the trie key with a relevant KeyFor* implementation to make sure the consistency

nearmax replied:
Please add this as a very large comment.

nearmax commented:
Is this ready to be merged?

frol commented:
I have addressed all the comments and finishing the test for Access Key changes, I will oush all of that in 3 hours, and it will be ready to be merged.

bowenwang1996 commented:
looks like it needs @evgenykuzyakov's approval as well.

nearmax commented:
> looks like it needs @evgenykuzyakov's approval as well.

I think it just needs the rebase

frol commented:
I addressed all the comments. It is fine to be merged.

evgenykuzyakov commented on core/primitives/src/types.rs:
```diff
@@ -116,20 +131,173 @@ pub enum StateChangeCause {
+
+#[derive(Debug)]
+pub enum StateChangeValue {
+    AccountUpdate { account_id: AccountId, account: Account },
```
Why `AccountUpdate` has `account_id` in it, but `AccessKeyUpdate` doesn't have `account_id`?

frol commented:
*The documentation request is https://github.com/nearprotocol/docs/issues/173*
